src/Controllers/PoolProviderController.cs

```diff
@@ -95,46 +96,66 @@ private void LogRequestBody()
+                (string orchestrationId, string jobName) = ExtractRequestSourceInfo();
 
-                string queueId;
-                try
+                using (_logger.BeginScope("Starting acquire operation for " +
+                    "Agent Id={agentId} Pool={agentPool} OrchestrationId={orchestrationId} JobName={jobName}",
+                    agentRequestItem.agentId, agentRequestItem.agentPool, orchestrationId, jobName))
                 {
-                    queueId = ExtractQueueId(agentRequestItem.agentSpecification);
+                    // To acquire a new agent, we'll need to do the following:
+                    // 1. Determine what queue VSTS is asking for.
+                    // 2. Determine whether such a queue exists in Helix (and whether we can use it)
+                    // 3. If the queue exists, submit work to it.
+                    string queueId;
+                    try
+                    {
+                        queueId = ExtractQueueId(agentRequestItem.agentSpecification);
 
-                    if (queueId == null)
+                        if (queueId == null)
+                        {
+                            return Json(new AgentInfoItem() { accepted = false });
+                        }
+                    }
+                    catch (Exception e)
                     {
-                        return Json(new AgentInfoItem() { accepted = false });
+                        _logger.LogError($"Unable to extract queue id from request: {e.ToString()}");
+                        return BadRequest();
                     }
-                }
-                catch (Exception e)
-                {
-                    _logger.LogError($"Unable to extract queue id from request: {e.ToString()}");
-                    return BadRequest();
-                }
 
-                var agentSettings = agentRequestItem.agentConfiguration.agentSettings;
+                    var agentSettings = agentRequestItem.agentConfiguration.agentSettings;
 
-                _logger.LogInformation("Acquiring agent for queue {queueId}", queueId);
+                    _logger.LogInformation("Acquiring agent for queue {queueId}", queueId);
 
-                var jobCreator = await GetHelixJobCreator(agentRequestItem, queueId, orchestrationId, jobName);
-                if (jobCreator == null)
-                {
-                    return Json(new AgentInfoItem() { accepted = false });
+                    var jobCreator = await GetHelixJobCreator(agentRequestItem, queueId, orchestrationId, jobName);
+                    if (jobCreator == null)
+                    {
+                        return Json(new AgentInfoItem() { accepted = false });
+                    }
+                    cancelTokenSource.Token.ThrowIfCancellationRequested();
+                    var agentInfoItem = await jobCreator.CreateJob(cancelTokenSource.Token);
+                    ForceGarbageCollection();
```

Comment:
Alex will have a PR for this soon (though it's in arcade). Can we wait for that and not have to mess around with the GC here?

Reply:
This also likely won't solve the problem. This usually happens when something is wrong with the connection, it only leaks when it fails. This code being _after_ the leaky things means that in the failure cases (when things leak and we are broken) we won't run the cleanup that will fix things, leaving it broken.

We should do the GC stuff before the call to CreateJob to make sure what we are about to do won't fail.

Reply:
Oh! I didn't the the caller above, ignore my previous comment. (Though I'd still like to not have this hacked checked in forever, it will slow us down)

Reply:
I'll revert the GC part if we get that rolled out tomorrow, we may as well still honor the cancellations.